sssd_test_framework/utils/sbus.py

```diff
@@ -0,0 +1,443 @@
+
+    def __setattr__(self, attr: str, value: Any) -> None:
+        """
+        If the caller sets a property, invoke the set_value() method
+        of the corresponding proxy property.
+        """
+        # This check avoids extra and recursive calls during object initialization
+        if "_properties" in self.__dict__ and attr in self._properties:
+            return self._properties[attr].set_value(value)
+
+        return super().__setattr__(attr, value)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(*args):
+        pass
+
+
+class DBUSDestination(MultihostUtility[MultihostHost]):
+    def __init__(self, host: MultihostHost, dest: str, bus: str):
+        """
+        Create a destination object associated to a bus and a name.
+
+        :param host: The host where the D-Bus serices run.
+        :type host: MultihostHost
+        :param dest: Destination application to contact.
+        :type dest: str
+        :param bus: The bus to use for the communications. Defaults to ``DBUSBus.SYSTEM``.
+            For other cases, a string can be provided with the explicit bus path.
+        :type bus: str | DBUSBus, optional
+        """
+        super().__init__(host)
+        self._dest = dest
+        self._bus = bus
+        self._objPaths: list[str] | None = None
+
+    def _instrospect(self, objPath: str) -> xml.Element:
+        introspection = ProxyMethod(self.host, self._dest, objPath, self._bus)
+        res = introspection._run("org.freedesktop.DBus.Introspectable.Introspect")
+        if res.rc != 0:
+            raise RuntimeError(f"Introspection failed for {objPath}")
+
+        xmlstr = res.stdout.removeprefix('   string "').removesuffix('"')
+        return xml.fromstring(xmlstr)
+
+    def _listNodes(self, path: str) -> list[str]:
+        lst: list[str] = []
+
+        root = self._instrospect(path)
+        if root.tag == "node":
+            lst.append(path)
+            for child in root:
+                if child.tag == "node":
+                    childPath = (path if path != "/" else "") + "/" + child.attrib["name"]
+                    lst += self._listNodes(childPath)
+
+        return lst
+
+    def getObjectPaths(self) -> list[str]:
+        """
+        Returns the list of the paths available at this destination.
+        """
+        if self._objPaths is None:
+            self._objPaths = self._listNodes("/")
+
+        return self._objPaths
+
+    def getObject(self, objpath: str) -> DBUSObject:
+        """
+        Creates and returns an object representing the D-Bus object and *object path*
+        associated to the *destination*.
+
+        :param objpath: The path to the object at destination.
+        :type objpath: str
+        """
+        return DBUSObject(self.host, dest=self._dest, objpath=objpath, bus=self._bus)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(*args):
+        pass
```

Comment:
I have no problem in removing it, but It was a requirement from you to make it work with the `with` statement.

Just let me know what you want me to do.

Reply:
I anticipated that we will maintain connection in some way, then it would make sense. But since we rely on dbus-send and don't maintain any resources on our own, we don't need it.